pkg/apis/workflow/v1alpha1/workflow_types.go

```diff
@@ -746,6 +764,47 @@ func (a *ArtifactLocation) HasLocation() bool {
 		a.GCS.HasLocation()
 }
 
+func (a *ArtifactLocation) GetType() (ArtifactLocationType, error) {
```

Comment:
minor - I'd actually return nil or `ArtifactLocationUnknown` for error case

Reply:
I'm more keen on returning `nil`. Figured I'd put an error because Golang runs on `if err != nil` 😄

Reply:
ok, zero value ("" I think, not nil) would be fine